dao/mongodb.rya/src/main/java/mvm/rya/mongodb/MongoConnectorFactory.java

```diff
@@ -41,26 +47,77 @@
-            final int port = Integer.parseInt(conf.get(MongoDBRdfConfiguration.MONGO_INSTANCE_PORT));
-            final ServerAddress server = new ServerAddress(host, port);
-
-            //check for authentication credentials
-            if (conf.get(MongoDBRdfConfiguration.MONGO_USER) != null) {
-                final String username = conf.get(MongoDBRdfConfiguration.MONGO_USER);
-                final String dbName = conf.get(MongoDBRdfConfiguration.MONGO_DB_NAME);
-                final char[] pswd = conf.get(MongoDBRdfConfiguration.MONGO_USER_PASSWORD).toCharArray();
-                final MongoCredential cred = MongoCredential.createCredential(username, dbName, pswd);
-                mongoClient = new MongoClient(server, Arrays.asList(cred));
+    public static synchronized MongoClient getMongoClient(final Configuration conf)
+            throws ConfigurationRuntimeException, MongoException {
+        if (mongoClient == null) {
+            // The singleton client has not yet created, is it a test/mock instance, or a service?
+            if (conf.getBoolean(MongoDBRdfConfiguration.USE_TEST_MONGO, false)) {
+                try {
+                    MongodForTestsFactory testsFactory = MongodForTestsFactory.with(Version.Main.PRODUCTION);
+                    mongoClient = testsFactory.newMongo();
+                } catch (IOException e) {
+                    // Rethrow as an unchecked error.  Since we are in a test mode here, just fail fast.
+                    throw new MongoException("While creating a factory for a test/mock MongoDB instance.",e);
+                }
             } else {
-                mongoClient = new MongoClient(server);
+                // Connect to a running Mongo server
+                final String host = requireNonNull(conf.get(MongoDBRdfConfiguration.MONGO_INSTANCE), "MongoDB host name is required");
+                final int port = requireNonNullInt(conf.get(MongoDBRdfConfiguration.MONGO_INSTANCE_PORT),
+                        "MongoDB Port number is required.");
+                ServerAddress server;
+                try {
+                    server = new ServerAddress(host, port);
+                } catch (UnknownHostException e) {
+                    throw new MongoException("Connecting MongoDB instance, cannot find host="+host,e);
+                }
+                // check for authentication credentials
+                if (conf.get(MongoDBRdfConfiguration.MONGO_USER) != null) {
+                    final String username = conf.get(MongoDBRdfConfiguration.MONGO_USER);
+                    final String dbName = requireNonNull(conf.get(MongoDBRdfConfiguration.MONGO_DB_NAME),
+                            MongoDBRdfConfiguration.MONGO_DB_NAME + " is null but required configuration if "
+                                    + MongoDBRdfConfiguration.MONGO_USER + " is configured.");
+                    final char[] pswd = requireNonNull(conf.get(MongoDBRdfConfiguration.MONGO_USER_PASSWORD),
```

Comment:
the requireNonNull() should be before this.  assignment should only happen after the precondition checks.